providers/implementations/keymgmt/hybrid_kmgmt.c

```diff
@@ -0,0 +1,668 @@
+#define GET_KEY_LENGTH(lname, type)                                         \
+    uint32_t ossl_hybrid_get_key_length_from_ ## lname(type *s)             \
+    {                                                                       \
+        if (s == NULL)                                                      \
+            return 0;                                                       \
+        if (s->key_lengths[MAX_HYBRID_ALGS] == 0)                           \
+            s->key_lengths[MAX_HYBRID_ALGS] =                               \
+                accumulate_ ## lname ## _uint32_params                      \
+                    (s->num_algs, s->lname ## s,                            \
+                     OSSL_PKEY_PARAM_MAX_SIZE, s->key_lengths);             \
+        return s->key_lengths[MAX_HYBRID_ALGS];                             \
+    }
+
+GET_KEY_LENGTH(key, HYBRID_PKEY)
+GET_KEY_LENGTH(ctx, HYBRID_PKEY_CTX)
+
+int ossl_hybrid_kmgmt_get_params(void *vkey, OSSL_PARAM params[])
+{
+    HYBRID_PKEY *key = (HYBRID_PKEY *)vkey;
+    OSSL_PARAM *p;
+    uint32_t t;
+    int type;
+
+    if (ossl_param_is_empty(params))
+        return 1;
+
+    for (p = params; params->key != NULL; params++) {
+        type = ossl_param_find_pidx(p->key);
+        switch (type) {
+        case PIDX_PKEY_PARAM_SECURITY_BITS:
+            t = maximum_key_uint32_params(key->num_algs, key->keys,
+                                          OSSL_PKEY_PARAM_SECURITY_BITS);
+            if (t == 0 || !OSSL_PARAM_set_uint32(p, t))
+                return 0;
+            break;
+
+        case PIDX_PKEY_PARAM_BITS:
+            t = accumulate_key_uint32_params(key->num_algs, key->keys,
+                                             OSSL_PKEY_PARAM_BITS, NULL);
+            if (t == 0 || !OSSL_PARAM_set_uint32(p, t))
+                return 0;
+            break;
+
+        case PIDX_PKEY_PARAM_MAX_SIZE:
+            t = ossl_hybrid_get_key_length_from_key(key);
+            if (t == 0 || !OSSL_PARAM_set_uint32(p, t))
+                return 0;
+            break;
+
+        default:
+            if (!key_subalg_get_param(key->num_algs, key->keys, p))
+                return 0;
+            break;
+        }
+    }
+    return 1;
+}
+
+const OSSL_PARAM *ossl_hybrid_kmgmt_gettable_params(void *vkey)
+{
+    HYBRID_PKEY *key = (HYBRID_PKEY *)vkey;
+    const OSSL_PARAM *pary[MAX_HYBRID_ALGS + 1];
+    unsigned int i;
+    static const OSSL_PARAM hybrid_gettable_params[] = {
+        OSSL_PARAM_int(OSSL_PKEY_PARAM_BITS, NULL),
+        OSSL_PARAM_int(OSSL_PKEY_PARAM_SECURITY_BITS, NULL),
+        OSSL_PARAM_int(OSSL_PKEY_PARAM_MAX_SIZE, NULL),
+        OSSL_PARAM_END
+    };
+
+    if (key->gettable == NULL) {
+        for (i = 0; i < key->num_algs; i++)
+            pary[i] = EVP_PKEY_gettable_params(key->keys[i]);
+        pary[i] = hybrid_gettable_params;
+        key->gettable = ossl_param_merge(key->num_algs + 1, pary);
```

Comment:
`EVP_PKEY`s can can be shared between threads since they are typically immutable once created. Looks like a potential for a race

Reply:
Another case of potential multiple writes of the same thing (albeit the pointers are different, what they point to will be equal).  Also thinking above these further.